From c605d3e9ec9c7f85cc80afdc6d44251943771bad Mon Sep 17 00:00:00 2001
From: Rafael Ostertag <rafisol@opencsw.org>
Date: Tue, 6 Sep 2011 21:32:16 +0200
Subject: [PATCH] Make glib work with zoneinfo version 1

Glib expect zoneinfo version 2, Solaris uses zoneinfo version 1.

See also

 * http://cs.ucla.edu/~eggert/tz/tz-link.htm
 * tzfile.h in ftp://elsie.nci.nih.gov/pub/tzcode2011i.tar.gz
 * /usr/include/tzfile.h
---
 glib/gtimezone.c |   39 ++++++++++++++++++++++++++++++++++++++-
 1 files changed, 38 insertions(+), 1 deletions(-)

diff --git a/glib/gtimezone.c b/glib/gtimezone.c
index e882682..6ef278e 100644
--- a/glib/gtimezone.c
+++ b/glib/gtimezone.c
@@ -121,7 +121,11 @@ struct _GTimeZone
 
   const struct tzhead *header;
   const struct ttinfo *infos;
+#ifndef __sun
   const gint64_be     *trans;
+#else
+  const gint32_be     *trans;
+#endif
   const guint8        *indices;
   const gchar         *abbrs;
   gint                 timecnt;
@@ -269,13 +273,23 @@ parse_constant_offset (const gchar *name,
 static GBuffer *
 zone_for_constant_offset (const gchar *name)
 {
+#ifndef __sun
   const gchar fake_zoneinfo_headers[] =
     "TZif" "2..." "...." "...." "...."
     "\0\0\0\0" "\0\0\0\0" "\0\0\0\0" "\0\0\0\0" "\0\0\0\0" "\0\0\0\0"
     "TZif" "2..." "...." "...." "...."
     "\0\0\0\0" "\0\0\0\0" "\0\0\0\0" "\0\0\0\0" "\0\0\0\1" "\0\0\0\7";
+#else
+  const gchar fake_zoneinfo_headers[] =
+    "TZif" "\0..." "...." "...." "...."
+    "\0\0\0\0" "\0\0\0\0" "\0\0\0\0" "\0\0\0\0" "\0\0\0\1" "\0\0\0\7";
+#endif
   struct {
+#ifndef __sun
     struct tzhead headers[2];
+#else
+    struct tzhead headers[1];
+#endif
     struct ttinfo info;
     gchar abbr[8];
   } *fake;
@@ -367,7 +381,11 @@ g_time_zone_new (const gchar *identifier)
 
               tzdir = getenv ("TZDIR");
               if (tzdir == NULL)
+#ifndef __sun
                 tzdir = "/usr/share/zoneinfo";
+#else
+	        tzdir = "/usr/share/lib/zoneinfo";
+#endif
 
               filename = g_build_filename (tzdir, identifier, NULL);
             }
@@ -383,8 +401,15 @@ g_time_zone_new (const gchar *identifier)
           const struct tzhead *header = tz->zoneinfo->data;
           gsize size = tz->zoneinfo->size;
 
-          /* we only bother to support version 2 */
+          /* we only bother to support version 2
+	   *
+	   * Well, I don't -- raos
+	   */
+#ifndef __sun
           if (size < sizeof (struct tzhead) || memcmp (header, "TZif2", 5))
+#else
+          if (size < sizeof (struct tzhead) || memcmp (header, "TZif\0", 5))
+#endif
             {
               g_buffer_unref (tz->zoneinfo);
               tz->zoneinfo = NULL;
@@ -393,6 +418,7 @@ g_time_zone_new (const gchar *identifier)
             {
               gint typecnt;
 
+#ifndef __sun
               /* we trust the file completely. */
               tz->header = (const struct tzhead *)
                 (((const gchar *) (header + 1)) +
@@ -402,6 +428,9 @@ g_time_zone_new (const gchar *identifier)
                   5 * guint32_from_be(header->tzh_timecnt) +
                   6 * guint32_from_be(header->tzh_typecnt) +
                   guint32_from_be(header->tzh_charcnt));
+#else
+	      tz->header = header;
+#endif
 
               typecnt     = guint32_from_be (tz->header->tzh_typecnt);
               tz->timecnt = guint32_from_be (tz->header->tzh_timecnt);
@@ -480,7 +509,11 @@ interval_start (GTimeZone *tz,
                 gint       interval)
 {
   if (interval)
+#ifndef __sun
     return gint64_from_be (tz->trans[interval - 1]);
+#else
+    return (gint64)gint32_from_be (tz->trans[interval - 1]);
+#endif
 
   return G_MININT64;
 }
@@ -490,7 +523,11 @@ interval_end (GTimeZone *tz,
               gint       interval)
 {
   if (interval < tz->timecnt)
+#ifndef __sun
     return gint64_from_be (tz->trans[interval]) - 1;
+#else
+    return (gint64)gint32_from_be (tz->trans[interval]) - 1;
+#endif
 
   return G_MAXINT64;
 }
-- 
1.7.6

